prov/efa/src/efa_cq.c

```diff
@@ -0,0 +1,535 @@
+ * @param[in]	err			positive libfabric error code
+ * @param[in]	prov_errno	positive EFA provider specific error code
+ */
+static void efa_cq_handle_error(struct efa_base_ep *base_ep, struct ibv_cq_ex *ibv_cq_ex, struct util_cq *util_cq, int err, int prov_errno) {
+	struct fi_cq_err_entry err_entry;
+	fi_addr_t addr;
+	int write_cq_err;
+
+	memset(&err_entry, 0, sizeof(err_entry));
+
+	err_entry.err = err;
+	err_entry.prov_errno = prov_errno;
+	err_entry.op_context = (void *)(uintptr_t)ibv_cq_ex->wr_id;
+	err_entry.flags = efa_cq_opcode_to_fi_flags(ibv_wc_read_opcode(ibv_cq_ex));
+	err_entry.buf = NULL;
+	err_entry.data = 0;
+	err_entry.tag = 0;
+
+	if (ibv_wc_read_wc_flags(ibv_cq_ex) & IBV_WC_WITH_IMM)
+		err_entry.data = ibv_wc_read_imm_data(ibv_cq_ex);
+
+	addr = efa_av_reverse_lookup(base_ep->av,
+				     ibv_wc_read_slid(ibv_cq_ex),
+				     ibv_wc_read_src_qp(ibv_cq_ex));
+
+	if (OFI_UNLIKELY(efa_write_error_msg(base_ep, addr, prov_errno,
+					     &err_entry.err_data,
+					     &err_entry.err_data_size)))
+		err_entry.err_data_size = 0;
+
+	EFA_WARN(FI_LOG_CQ, "err: %d, message: %s (%d)\n",
+		err_entry.err,
+		err_entry.err_data
+			? (const char *) err_entry.err_data
+			: efa_strerror(err_entry.prov_errno),
+		err_entry.prov_errno);
+
+	efa_show_help(err_entry.prov_errno);
+
+	efa_cntr_report_error(&base_ep->util_ep, err_entry.flags);
+	write_cq_err = ofi_cq_write_error(util_cq, &err_entry);
+	if (write_cq_err) {
+		EFA_WARN(FI_LOG_CQ,
+			"Error writing error cq entry when handling TX error\n");
+		efa_base_ep_write_eq_error(base_ep, err, prov_errno);
+	}
+}
+
+
+/**
+ * @brief handle the event that a send request has been completed
+ *
+ * @param[in]		ibv_cq_ex	extended ibv cq
+ * @param[in]		base_ep	    efa_base_ep
+ * @param[in]       cq_entry    fi_cq_tagged_entry
+ */
+static void efa_cq_handle_send_completion(struct efa_base_ep *base_ep,
+					  struct ibv_cq_ex *ibv_cq_ex,
+					  struct fi_cq_tagged_entry *cq_entry)
+{
+	struct util_cq *tx_cq = base_ep->util_ep.tx_cq;
+	int ret = 0;
+
+	/* wr_id = NULL means no completion */
+	if (!ibv_cq_ex->wr_id)
+		return;
+
+	if (base_ep->util_ep.tx_msg_flags == FI_COMPLETION ||
+	    cq_entry->flags & FI_COMPLETION) {
+		/* TX completions should not send peer address to util_cq */
+		if (base_ep->util_ep.caps & FI_SOURCE) {
+			ret = ofi_cq_write_src(tx_cq, cq_entry->op_context,
+					       cq_entry->flags, cq_entry->len,
+					       cq_entry->buf, cq_entry->data,
+					       cq_entry->tag, FI_ADDR_NOTAVAIL);
+		} else {
+			ret = ofi_cq_write(tx_cq, cq_entry->op_context,
+					   cq_entry->flags, cq_entry->len,
+					   cq_entry->buf, cq_entry->data,
+					   cq_entry->tag);
+		}
+
+		if (OFI_UNLIKELY(ret)) {
+			EFA_WARN(FI_LOG_CQ,
+				"Unable to write send completion: %s\n",
+				fi_strerror(-ret));
+			efa_cq_handle_error(base_ep, ibv_cq_ex, tx_cq, -ret,
+					    FI_EFA_ERR_WRITE_SEND_COMP);
+			return;
+		}
+	}
+
+	if (cq_entry->flags != 0)
```

Comment:
IIRC cntr should always be updated for any completion. In what cases you will have a `cq_entry->flags` as 0?

Reply:
fabtests use FI_CQ_FORMAT_CONTEXT, which doesn't have flags(see `efa_cq_read_context_entry`)

Reply:
that flags can be derived from op code, it doesn't depend on format.